In [209]:
# get individual worm data
# Conny Lin | June 1, 2020

In [1]:
import numpy as np
import pandas as pd
import os
from pathlib import Path
from pathlib import PurePath
import pathlib
import glob
import pickle

In [2]:
dir_save = '/Users/connylin/Dropbox/CA/ED _20200119 Brain Station Data Science Diploma/Capstone/data'

# Search DB for plates

In [3]:
# load file summary
filename = os.path.join(dir_save, 'file_summary_mwt.pickle')
with open(filename,'rb') as f:
    file_summary = pickle.load(f)

In [4]:
file_summary.shape

(7298, 38)

In [214]:
file_summary

filepath                    \
suffix                                                      .blob .blobs .csv   
mwtpath                                                                         
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...      NaN   NaN    2.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...      NaN   NaN    1.0  NaN   
...                                                     ...   ...    ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    5.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    7.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    3.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    3.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    1.0  NaN   

                                                                            \
suffix                                             .dat .db .eps .fig .jpg   
mwtpath                                                                      
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...  1.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...  2.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...  1.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...  NaN NaN  NaN  NaN  NaN   
...                                                 ...  ..  ...  ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   

                                                                           \
suffix                                              .m .mat  .p .pdf .png   
mwtpath                                                                     
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3... NaN  NaN NaN  NaN  1.0   
...                                                 ..  ...  ..  ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   

                                                                               \
suffix                                             .rasterData .rev .set .ssr   
mwtpath                                                                         
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...         NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...         NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...         NaN  1.0  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...         NaN  1.0  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...         NaN  NaN  1.0  NaN   
...                                                        ...  ...  ...  ...   
/Volumes/COBOLT

In [215]:
i = ~file_summary.iloc[:,-1].isna()
i.values

array([False, False, False, ..., False, False, False])

In [216]:
plates_with_trinity = file_summary.iloc[i.values,-1].copy()

In [217]:
plates_with_trinity.shape

(871,)

In [218]:
i = ~file_summary.iloc[:,-2].isna()
plates_with_swanlake = file_summary.iloc[i.values,-1].copy()
plates_with_swanlake.shape

(30,)

More plates with trinity than swanlake

In [219]:
plates_with_trinity.head()

mwtpath
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_113945    408.0
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_121214    495.0
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_124444    406.0
/Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s/NM1630/20110907_102613      369.0
/Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s/NM1630/20110907_103523      427.0
Name: (filepath, trinity.id.dat), dtype: float64

In [220]:
# get mwt paths
mwtpaths = pd.DataFrame(plates_with_trinity.index.values, columns=['paths'])
mwtpaths

paths
0    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
1    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
2    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
3    /Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s...
4    /Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s...
..                                                 ...
866  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
867  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
868  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
869  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
870  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...

[871 rows x 1 columns]

In [221]:
# get which ones are alcohol and which ones are not
# split paths 
# MWTDB = pd.DataFrame()
MWTDB = mwtpaths['paths'].str.split(pat='/', expand=True)
MWTDB

0         1       2    3                                  4             5  \
0       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
1       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
2       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
3       Volumes  COBOLT  MWT         20110907B_SS_100s30x10s10s        NM1630   
4       Volumes  COBOLT  MWT         20110907B_SS_100s30x10s10s        NM1630   
..  ..      ...     ...  ...                                ...           ...   
866     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
867     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
868     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
869     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
870     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   

                   6  
0    20110906_113945  
1    20110906_121214  
2    20110906_124444  
3    20110907_102613  
4    20110907_103523  
..               ...  
866  20170605_134308  
867  20170605_141815  
868  20170605_144356  
869  20170605_152520  
870  20170605_154338  

[871 rows x 7 columns]

In [222]:
# see unique groups
pd.set_option('display.max_rows',100)
pd.DataFrame(MWTDB.iloc[:,5].value_counts().sort_values(ascending=False))

5
N2                            189
N2_400mM                      146
NM1630                         40
NM1968                         34
VC228                          18
VC228_400mM                    18
KP4                            16
KP4_400mM                      14
TM774_400mM                    12
VM3109_400mM                   12
VM3109                         12
NM1630_400mM                   12
N2_400mM_ExpTime60min          11
RB1231_400mM                   10
N2_400mM_NoFood                 9
BZ142                           9
TM5182_400mM                    8
TM2659                          8
TM474_400mM                     8
TM2179                          8
N2_400mM_ExpTime120min          8
TM1630                          8
TM474                           8
TM1630_400mM                    8
RM3389                          8
TM5182                          8
N2_400mM_ExpTime0min            8
RM3389_400mM                    8
TM2659_400mM                    8
N2_NoFood                       8
TM774                           7
DG1856                          7
TM8625                          7
TM2179_400mM                    7
TM8625_400mM                    7
JT748                           6
KP1097                          6
VC40862                         6
JT748_400mM                     6
VC40516                         6
VC40516_400mM                   6
VC40862_400mM                   6
PS2627                          6
N2_E3d12h400mM_R12h_T4d400mM    6
KP1097_400mM                    6
TM4724                          6
PS2627_400mM                    6
TM4724_400mM                    6
NM1968_Paxillin1pM              5
NM1968_Paxillin10pM             5
N2_Paxillin100pM                5
N2_400mM_ExpTime90min           4
RB989_400mM                     4
VC1416                          4
RB989                           4
N2_TsfLiquidTime120to180        4
N2_TsfLiquidTime30to90          4
VG302_400mM                     4
VG302                           4
VC1416_400mM                    4
SG1_400mM                       4
CX3940                          4
SG1                             4
DA465                           3
NM1968_Paxillin1mM              3
NM1968_Paxillin10mM             3
CX3940_Paxillin100pM            2
NM1968_Paxillin10uM             2
NM1968_Paxillin1uM              2
NM1968_Paxillin100pM            2
CX3940_Paxillin1pM              1
RB982_400mM                     1
VG202_400mM_NoLid               1
MT2426                          1

In [223]:
# take just N2 and N2 400mM to train
i = MWTDB.iloc[:,5].str.match(pat='^N2$') | MWTDB.iloc[:,5].str.match(pat='^N2_400mM$')
# check to see if plate names are unique
if MWTDB.loc[i,6].unique().shape == np.sum(i):
    print('plate names unique')
mwtpaths_target = MWTDB.loc[i,:].copy()
mwtpaths_target['mwtpath'] = mwtpaths[i]


plate names unique


# get paths to trinity data from file database


In [224]:
#load file database csv
df = pd.read_csv(os.path.join(dir_save,'path_allfiles_MWTDB.csv'))
df

Unnamed: 0        dbname                        expname groupname  \
0                 0           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
1                 1           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
2                 2           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
3                 3           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
4                 4           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
...             ...           ...                            ...       ...   
1318719     1318719  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318720     1318720  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318721     1318721  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318722     1318722  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318723     1318723  MWT to check     20140819C_SM_100s30x10s10s   ungroup   

                 mwtname                                           filename  \
0        20100201_131356                                        0%_EtOH.dat   
1        20100201_131356                                        0%_EtOH.png   
2        20100201_131356                                    0%_EtOH.summary   
3        20100201_131356                                   0%_EtOH.evan.dat   
4        20100201_131356                          0%_EtOH.drunkposture2.dat   
...                  ...                                                ...   
1318719  20140819_192732           N2_5x3_t96h20C_100s30x10s10s_C0109dd.set   
1318720  20140819_192732       N2_5x3_t96h20C_100s30x10s10s_C0109dd.summary   
1318721  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd_00000k.blobs   
1318722  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd_00001k.blobs   
1318723  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd.drunkpost...   

                                                      path  \
0        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
1        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
2        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
3        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
4        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
...                                                    ...   
1318719  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318720  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318721  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318722  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318723  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   

                                                     fname       ext  \
0                                                  0%_EtOH      .dat   
1                                                  0%_EtOH      .png   
2                                                  0%_EtOH  .summary   
3                                             0%_EtOH.evan      .dat   
4                                    0%_EtOH.drunkposture2      .dat   
...                                                    ...       ...   
1318719               N2_5x3_t96h20C_100s30x10s10s_C0109dd      .set   
1318720               N2_5x3_t96h20C_100s30x10s10s_C0109dd  .summary   
1318721        N2_5x3_t96h20C_100s30x10s10s_C0109dd_00000k    .blobs   
1318722        N2_5x3_t96h20C_100s30x10s10s_C0109dd_00001k    .blobs   
1318723  N2_5x3_t96h20C_100s30x10s10s_C0109dd.drunkpost...      .dat   

               analysis_tag  
0                        []  
1                        []  
2                        []  
3                 ['.evan']  
4        ['.drunkposture2']  
...                     ...  
1318719                  []  
1318720                  []  
1318721                  []  
1318722                  []  
1318723  ['.drunkposture2']  

[1318724 rows x 10 columns]

In [225]:
i = df['mwtname'].isin(mwtpaths_target.iloc[:,6].values)


In [226]:
sum(i)/len(i)

0.1718486961638675

In [227]:
# reduce data frame
df = df.loc[i,:].copy()
df.shape

(226621, 10)

In [228]:
i = df['filename'].str.contains(r'trinity[.]\d{1,}(.)dat$')

/Users/connylin/opt/anaconda3/lib/python3.7/site-packages/pandas/core/strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


In [229]:
# reduce again to only file name containing trinity
df = df.loc[i,:].copy()
df.shape

(216261, 10)

In [230]:
# print some to confirm
for f in df['filename'].iloc[:10]:
    print(f)

N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00001.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00002.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04098.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00004.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00005.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04101.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00006.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04103.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00008.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00009.dat


In [231]:
# group by 0mM and 400mM to see how many data we have per group
df['groupname'].value_counts()

N2             120524
N2_400mM        92886
TM774_400mM      2799
VC40862            52
Name: groupname, dtype: int64

In [232]:
# found some dirty data. delete those
i = df['groupname'].isin(['TM774_400mM','VC40862'])
df.drop(df.index[i], axis=0, inplace=True)


In [233]:
df.shape

(213410, 10)

In [234]:
# group by 0mM and 400mM to see how many data we have per group
df['groupname'].value_counts()

N2          120524
N2_400mM     92886
Name: groupname, dtype: int64

Now the db is clean

the samples are a little unbalanced but not too bad.

In [235]:
# store df in more permanent variable name
fileDB_target = df.copy()
del df

In [269]:
# get worm id
worm_number = fileDB_target['fname'].str.extract(r'(?<=trinity[.])(\d{1,})')
worm_number = worm_number.astype('int32').values

In [334]:
fileDB_target['worm_id'] = worm_number

# take a look at trinity data

In [237]:
# sample data to see how to import the data
pfile = fileDB_target['path'].iloc[200]
pfile

'/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_113945/N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.05483.dat'

In [238]:
df = pd.read_csv(pfile, delim_whitespace=True, header=None)
df
# if the file has worm id can also concat directly from bash

0    1    2       3    4    5    6    7        8        9       10  \
0    1709.275  1.0  1.0     NaN  0.0  1.0  0.0  0.0  39.0142  18.2930  0.1093   
1    1709.360  1.0  1.0  0.1072  0.0  1.0  0.0  0.0  39.0191  18.3024  0.1112   
2    1709.446  1.0  1.0  0.0959  0.0  1.0  0.0  0.0  39.0225  18.3093  0.1094   
3    1709.518  1.0  1.0  0.0956  0.0  1.0  0.0  0.0  39.0256  18.3159  0.1106   
4    1709.592  1.0  1.0  0.0713  0.0  1.0  0.0  0.0  39.0286  18.3218  0.1120   
..        ...  ...  ...     ...  ...  ...  ...  ...      ...      ...     ...   
489  1739.672  1.0  1.0  0.1655  0.0  1.0  0.0  0.0  40.4880  20.4281  0.1121   
490  1739.711  1.0  1.0  0.2041  0.0  0.0  0.0  0.0  40.4941  20.4343  0.1141   
491  1739.748  1.0  1.0  0.2142  0.0  0.0  0.0  0.0  40.5004  20.4388  0.1144   
492  1739.811  1.0  1.0     NaN  0.0  0.0  0.0  0.0  40.5106  20.4479  0.1130   
493  1739.877  1.0  1.0     NaN  0.0  0.0  0.0  0.0  40.5225  20.4573  0.1108   

         11        12   13     14    15    16      17  
0    1.2487  0.173502  NaN  0.208  20.1  35.5     NaN  
1    1.2643  0.175689  5.2  0.202  19.4  37.2  0.0012  
2    1.2437  0.174231  4.5  0.207  22.4  39.0  0.0030  
3    1.2483  0.174231  4.6  0.201  23.3  38.9  0.0017  
4    1.2594  0.175689  4.0  0.199  22.8  38.6  0.0054  
..      ...       ...  ...    ...   ...   ...     ...  
489  1.3017  0.177147  9.1  0.365  43.5  38.5  0.0289  
490  1.2576  0.175689  7.4  0.359  51.0  40.9  0.0312  
491  1.2671  0.176418  6.6  0.330  46.7  41.1  0.0301  
492  1.2656  0.174960  NaN  0.337  46.6  41.4     NaN  
493  1.2987  0.175689  NaN  0.325  49.6  41.6     NaN  

[494 rows x 18 columns]

In [239]:
# load file summary
filename = 'legend_trinity_worm.pickle'
with open(os.path.join(dir_save, filename),'rb') as f:
    v = pickle.load(f)

In [245]:
df.columns = v['name'].values

In [250]:
# only keep variables that are informative
col_ind_keep = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]
df = pd.read_csv(pfile, usecols=col_ind_keep, delim_whitespace=True, 
                 header=None, names=v.loc[col_ind_keep,'name'].values)

time   speed  bias  tap    loc_x    loc_y  morphwidth  midline  \
0    1709.275     NaN   1.0  0.0  39.0142  18.2930      0.1093   1.2487   
1    1709.360  0.1072   1.0  0.0  39.0191  18.3024      0.1112   1.2643   
2    1709.446  0.0959   1.0  0.0  39.0225  18.3093      0.1094   1.2437   
3    1709.518  0.0956   1.0  0.0  39.0256  18.3159      0.1106   1.2483   
4    1709.592  0.0713   1.0  0.0  39.0286  18.3218      0.1120   1.2594   
..        ...     ...   ...  ...      ...      ...         ...      ...   
489  1739.672  0.1655   1.0  0.0  40.4880  20.4281      0.1121   1.3017   
490  1739.711  0.2041   0.0  0.0  40.4941  20.4343      0.1141   1.2576   
491  1739.748  0.2142   0.0  0.0  40.5004  20.4388      0.1144   1.2671   
492  1739.811     NaN   0.0  0.0  40.5106  20.4479      0.1130   1.2656   
493  1739.877     NaN   0.0  0.0  40.5225  20.4573      0.1108   1.2987   

         area  angular  aspect  kink  curve    crab  
0    0.173502      NaN   0.208  20.1   35.5     NaN  
1    0.175689      5.2   0.202  19.4   37.2  0.0012  
2    0.174231      4.5   0.207  22.4   39.0  0.0030  
3    0.174231      4.6   0.201  23.3   38.9  0.0017  
4    0.175689      4.0   0.199  22.8   38.6  0.0054  
..        ...      ...     ...   ...    ...     ...  
489  0.177147      9.1   0.365  43.5   38.5  0.0289  
490  0.175689      7.4   0.359  51.0   40.9  0.0312  
491  0.176418      6.6   0.330  46.7   41.1  0.0301  
492  0.174960      NaN   0.337  46.6   41.4     NaN  
493  0.175689      NaN   0.325  49.6   41.6     NaN  

[494 rows x 14 columns]

## How to effectively concat worm files

Research:
* https://stackoverflow.com/questions/45091398/pandas-combine-multiple-csvs-and-output-as-one-large-file

In [ ]:
# count rows for each file
# create worm id with that many rows

In [270]:
fileDB_target.shape

(213410, 10)

In [332]:
# get row numbers for all of the files
# declare empty array to store the row numbers
wormid_array = np.empty((fileDB_target.shape[0],1),dtype='int')
# loop through each files 
for i, p in enumerate(fileDB_target['path'], 0):
    # report process every 100 files
    if (i%100==0):
        print(f'processing #{i} files', end='\r')
    # create custom bash command
    bashCommand = "cat "+p+" | wc -l"
    # call bash
    output = subprocess.check_output(bashCommand, shell=True)
    # take out digits from the byte output
    wormid_array[i] = np.array(output,dtype='int')
wormid_array

array([[ 527],
       [ 915],
       [9648],
       ...,
       [ 155],
       [1649],
       [ 714]])

In [335]:
fileDB_target['row_number'] = wormid_array

In [337]:
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  
39135    ['.00001' '.trinity']        1         527  
39136    ['.00002' '.trinity']        2         915  
39137    ['.04098' '.trinity']     4098        9648  
39138    ['.00004' '.trinity']        4        1611  
39139    ['.00005' '.trinity']        5         707  
...                        ...      ...         ...  
1307197  ['.02552' '.trinity']     2552        1186  
1307198  ['.00505' '.trinity']      505         477  
1307199  ['.02041' '.trinity']     2041         155  
1307200  ['.01531' '.trinity']     1531        1649  
130

In [336]:
# save in pickle
filename = 'fileinfo_trinity_N2400mM.pickle'
with open(os.path.join(dir_save, filename),'wb') as f:
    pickle.dump(fileDB_target,f)

https://stackoverflow.com/questions/20716812/saving-and-loading-multiple-objects-in-pickle-file
```
import pickle

file = open('test.pkl','wb')
obj_1 = ['test_1', {'ability', 'mobility'}]
obj_2 = ['test_2', {'ability', 'mobility'}]
obj_3 = ['test_3', {'ability', 'mobility'}]

pickle.dump(obj_1, file)
pickle.dump(obj_2, file)
pickle.dump(obj_3, file)

file.close()

file = open('test.pkl', 'rb')
obj_1 = pickle.load(file)
obj_2 = pickle.load(file)
obj_3 = pickle.load(file)
print(obj_1)
print(obj_2)
print(obj_3)
file.close()
```

In [11]:
# load pickle
# save in pickle
filename = 'fileinfo_trinity_N2400mM.pickle'
fileDB_target = pickle.load(open(os.path.join(dir_save, filename),'rb'))
    

In [12]:
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  
39135    ['.00001' '.trinity']        1         527  
39136    ['.00002' '.trinity']        2         915  
39137    ['.04098' '.trinity']     4098        9648  
39138    ['.00004' '.trinity']        4        1611  
39139    ['.00005' '.trinity']        5         707  
...                        ...      ...         ...  
1307197  ['.02552' '.trinity']     2552        1186  
1307198  ['.00505' '.trinity']      505         477  
1307199  ['.02041' '.trinity']     2041         155  
1307200  ['.01531' '.trinity']     1531        1649  
130

In [6]:
# look at estimates
fileDB_target['row_number'].sum()

202718463

202,718,463. 202 Million rows.

# Take 1000 worms from each group to test alogorithms

In [32]:
# get sample
n_sample = 10
groupnames = ['N2','N2_400mM']
groupname_target = 'N2'

In [33]:
# separate by group
i = fileDB_target['groupname']==groupname_target
df_target = fileDB_target.iloc[i.values,:].copy()

In [34]:
df_target.shape

(120524, 12)

In [ ]:
df_sample = df_target.sample(n_sample, random_state=318).copy()

In [49]:
# get row total
row_total = df_sample['row_number'].sum()
# declare array
worm_id_array = np.empty([row_total,1],dtype='int')
worm_id_list = df_sample['worm_id'].values
rows_per_worm_list = df_sample['row_number'].values
i_row_previous = 0
for worm_id, rows in zip(worm_id_list, rows_per_worm_list):
    # get start row position
    i_start_row = i_row_previous
    # get end row position
    i_end_row = i_row_previous+rows-1
    print(i_row_previous, worm_id, rows, i_start_row, i_end_row)
    # get worm id array x number of rows
    worm_id_array[i_start_row:i_end_row+1,0] = np.tile(worm_id, rows)
    # create next starting position
    i_row_previous = rows+i_row_previous

0 12099 1630 0 1629
1630 3187 236 1630 1865
1866 17056 262 1866 2127
2128 402 267 2128 2394
2395 4610 798 2395 3192
3193 2895 788 3193 3980
3981 1327 257 3981 4237
4238 23505 113 4238 4350
4351 6735 470 4351 4820
4821 18168 1095 4821 5915


array([12099,  3187, 17056,   402,  4610,  2895,  1327, 23505,  6735,
       18168], dtype=int32)

https://www.datacamp.com/community/tutorials/shell-commands-data-scientist

In [ ]:
# concat all csv


In [28]:
# only keep variables that are informative
col_ind_keep = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]

df = pd.read_csv(pfile, usecols=col_ind_keep, delim_whitespace=True, 
                 header=None, 
                 names=v.loc[col_ind_keep,'name'].values)

array([ 3876,  9353,  6015, 12895, 22685,     7, 26232,  8225,  1325,
         354], dtype=int32)